### PR TITLE
Array field optional sortability

### PR DESCRIPTION
**Pull request by olzraiti**
In a complex form the sortability makes the form a bit cluttered and is often not needed.

I added a whole new object `ui:options` to the ArrayField's uiSchema for flexibility in case we add other options to ArrayField in the future.

### CODE REVIEW

File: test/ArrayField_test.js [hunk 1]
```diff
@@ -157,6 +157,15 @@ describe("ArrayField", () => {
       expect(moveDownBtns[1].disabled).eql(true);
     });
 
+    it("should not be sortable when uiSchema[\"ui:options\"].sortable === false", () => {
+      const {node} = createFormComponent({schema, formData: ["foo", "bar"], uiSchema: {"ui:options": {sortable: false}}});
+      const moveUpBtns = node.querySelector(".array-item-move-up");
+      const moveDownBtns = node.querySelector(".array-item-move-down");
+
+      expect(moveUpBtns).to.be.null;
+      expect(moveDownBtns).to.be.null;
```
Comment (leplatrem) [inline test/ArrayField_test.js]: 👍

File: test/ArrayField_test.js [hunk 2]
```diff
@@ -157,6 +157,15 @@ describe("ArrayField", () => {
       expect(moveDownBtns[1].disabled).eql(true);
     });
 
+    it("should not be sortable when uiSchema[\"ui:options\"].sortable === false", () => {
```
Comment (leplatrem) [inline test/ArrayField_test.js]: *should not show move up/down buttons is sortable is false*

File: README.md [hunk 3]
```diff
@@ -399,6 +400,24 @@ render((
         uiSchema={uiSchema} />
 ), document.getElementById("app"));
 ```
+### Array items ordering
+
+Array items are sortable by default. The `uiSchema` object spec allows you to disable ordering:
```
Comment (leplatrem) [inline README.md]: *are orderable by default, and react-jsonschema-form renders move up/down buttons along them.*

File: README.md [hunk 4]
```diff
@@ -399,6 +400,24 @@ render((
...
+  type: "array",
+  properties: {
+    type: "string"
+  }
+};
+
+const uiSchema = {
+  "ui:options":  {
+    sortable: false
+  }
```
Comment (leplatrem) [inline README.md]: I think that @n1k0 took a different approach in #346 in order to introduce widget options: 
```js
const uiSchema = {
  "ui:widget": {
    component: "checkboxes",
    options: {
      inline: true
    }
  }
};
```

I think that both approaches are valid :) I'll let you figure out pros and cons for each ;)

File: src/components/fields/ArrayField.js [hunk 5]
```diff
@@ -337,8 +337,15 @@ class ArrayField extends Component {
     autofocus
   }) {
     const {SchemaField} = this.props.registry.fields;
-    const {disabled, readonly} = this.props;
+    const {disabled, readonly, uiSchema} = this.props;
+
+    const sortable = uiSchema["ui:options"] ? uiSchema["ui:options"].sortable : true;
```
Comment (leplatrem) [inline src/components/fields/ArrayField.js]: What If options is defined but `sortable` is left undefined?

**Comment by olzraiti** [comment]
I renamed the option to 'orderable'.

File: README.md (hunk 4 above)
Comment (olzraiti) [inline README.md]: Having the same approach as widgets would be more consistent, but the `ui:options` -style makes updating a `uiSchema` simpler (if you want to add an option to an existing uiSchema field, you don't replace any lines, just add new ones).

File: src/components/fields/ArrayField.js [hunk 6]
```diff
@@ -337,8 +337,19 @@ class ArrayField extends Component {
...
   }) {
     const {SchemaField} = this.props.registry.fields;
-    const {disabled, readonly} = this.props;
+    const {disabled, readonly, uiSchema} = this.props;
+
+    let orderable = true;
+    const options = uiSchema["ui:options"];
+    if (options && options.hasOwnProperty("orderable")) {
+      orderable = !!options.orderable;
+    }
```
Comment (n1k0) [inline src/components/fields/ArrayField.js]: How about 

```js
const options = {orderable: true, ...uiSchema["ui:options"]}
```

File: src/components/fields/ArrayField.js [hunk 7]
```diff
@@ -337,8 +337,19 @@ class ArrayField extends Component {
...
+    if (!orderable) {
+      canMoveUp = false;
```
Comment (n1k0) [inline src/components/fields/ArrayField.js]: As a general good practice, I'd rather not mutate passed arguments at all.

File: src/components/fields/ArrayField.js [hunk 8]
```diff
@@ -337,8 +337,19 @@ class ArrayField extends Component {
...
+      canMoveDown = false;
+    }
     const hasToolbar = removable || canMoveUp || canMoveDown;
```
Comment (n1k0) [inline src/components/fields/ArrayField.js]: Actually the same could be achieved by this (I think):

```js
const {orderable} = {orderable: true, ...uiSchema["ui:options"]};
const hasToolbar = orderable && (removable || canMoveUp || canMoveDown);
```

File: README.md (hunk 3 above)
Comment (glasserc) [inline README.md]: I would probably say "alongside" :)

**Comment by n1k0** [comment]
Thank you! :+1:

**Comment by n1k0** [comment]
Released in [v0.41.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.41.0).

**Pull request merged**